src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs

```diff
@@ -0,0 +1,183 @@
+                private readonly PooledDictionary<Symbol, int> _symbolToSlot = PooledDictionary<Symbol, int>.GetInstance();
+                private readonly ArrayBuilder<int> _previousWalkerSlots = ArrayBuilder<int>.GetInstance();
+                private int _currentWalkerSlot = -1;
+                private int _nextWalkerSlot = 0;
+
+                internal SnapshotManager ToManagerAndFree()
+                {
+                    Debug.Assert(_currentWalkerSlot == -1 && _previousWalkerSlots.Count == 0, "Attempting to finalize snapshots before all walks completed");
+                    Debug.Assert(_walkerStates.Count == _nextWalkerSlot);
+                    Debug.Assert(_symbolToSlot.Count == _walkerStates.Count);
+                    _symbolToSlot.Free();
+                    _previousWalkerSlots.Free();
+                    return new SnapshotManager(
+                        _walkerStates.ToImmutableAndFree(),
+                        _incrementalSnapshots.ToImmutable());
+                }
+
+                internal void EnterNewWalker(Symbol symbol)
+                {
+                    Debug.Assert(!(symbol is null));
+                    _previousWalkerSlots.Push(_currentWalkerSlot);
+
+                    // Because we potentially run multiple passes, we
+                    // need to make sure we use the same final shared
+                    // state for following passes.
+                    if (_symbolToSlot.ContainsKey(symbol))
+                    {
+                        _currentWalkerSlot = _symbolToSlot[symbol];
+                    }
+                    else
+                    {
+                        _currentWalkerSlot = _nextWalkerSlot;
+                        _symbolToSlot.Add(symbol, _currentWalkerSlot);
+                        _nextWalkerSlot++;
+                    }
+                }
+
+                internal void ExitWalker(SharedWalkerState stableState)
+                {
+                    _walkerStates.SetItem(_currentWalkerSlot, stableState);
+                    _currentWalkerSlot = _previousWalkerSlots.Pop();
+                }
+
+                internal void TakeIncrementalSnapshot(BoundNode node, LocalState currentState)
+                {
+                    if (node == null || node.WasCompilerGenerated)
+                    {
+                        return;
+                    }
+
+                    // Note that we can't use Add here, as this is potentially not the stable
+                    // state of this node and we could get updated states later.
+                    _incrementalSnapshots[node.Syntax.SpanStart] = new Snapshot(currentState.Clone(), _currentWalkerSlot);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Contains the shared state state used to restore the walker at a specific point
+        /// </summary>
+        internal struct SharedWalkerState
```

Comment:
>SharedWalkerState [](start = 24, length = 17)

Is this different from the `VariableState` type?

Reply:
Yes. `VariableState` does not hold onto the `Symbol`, and it does hold onto a `LocalState`.

---
In reply to: [289521654](https://github.com/dotnet/roslyn/pull/35850#discussion_r289521654) [](ancestors = 289521654)